internal/plugin/connectors/http/generic/oauthprotocol/protocol.go

```diff
@@ -0,0 +1,214 @@
+type OAuth1 struct {
+	ConsumerKey     string
+	ConsumerSecret  string
+	Nonce           string
+	Signature       string
+	SignatureMethod string
+	TimeStamp       string
+	Token           string
+	TokenSecret     string
+	Version         string
+}
+
+// Oauth Keys - we set this as a constant to make sure that we always verify
+// this exact key syntax and spelling
+const (
+	oauth_consumer_key     = "oauth_consumer_key"
+	oauth_consumer_secret  = "oauth_consumer_secret"
+	oauth_token            = "oauth_token"
+	oauth_token_secret     = "oauth_token_secret"
+	oauth_nonce            = "oauth_nonce"
+	oauth_signature        = "oauth_signature"
+	oauth_signature_method = "oauth_signature_method"
+	oauth_version          = "oauth_version"
+	oauth_timestamp        = "oauth_timestamp"
+
+	oauth_version_number        = "1.0"
+	oauth_signature_method_type = "HMAC-SHA1"
+	nonce_length                = 16
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyz" +
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+	"0123456789"
+
+func stringWithCharset(length int, charset string) string {
+	var seededRand *rand.Rand = rand.New(
+		rand.NewSource(time.Now().UnixNano()))
+
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[seededRand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func generateNonce(length int) string {
+	return stringWithCharset(length, charset)
+}
+
+func checkRequiredOAuthParams(params map[string]string) error {
+	var requiredParams = []string{
+		oauth_consumer_key,
+		oauth_consumer_secret,
+		oauth_token,
+		oauth_token_secret,
+	}
+
+	for _, requiredParam := range requiredParams {
+		if param := params[requiredParam]; len(param) <= 0 {
+			return fmt.Errorf("required OAuth1 parameter '%s' not found", requiredParam)
+		}
+	}
+	return nil
+}
+
+// ExtractOAuthParams extracts the required OAuth1 parameters from the config
+// and assigns the other required paramters
+// If a required parameter is missing from the config throws an error
+func extractOAuthParams(params map[string]string) OAuth1 {
+	var oauth OAuth1
+
+	oauth.ConsumerKey = params[oauth_consumer_key]
+	oauth.ConsumerSecret = params[oauth_consumer_secret]
+	oauth.Token = params[oauth_token]
+	oauth.TokenSecret = params[oauth_token_secret]
+	oauth.TimeStamp = strconv.Itoa(int(time.Now().Unix()))
+	oauth.Version = oauth_version_number
+	oauth.SignatureMethod = oauth_signature_method_type
+	oauth.Nonce = generateNonce(nonce_length)
+
+	return oauth
+}
+
+func encodeURI(stringToEncode string) string {
+	return strings.ReplaceAll(url.QueryEscape(stringToEncode), "+", "%20")
+}
+
+// Create a Parameter String from the required OAuth params and query & body params
+// https://tools.ietf.org/html/rfc5849#section-3.5
+func generateParameterString(oauth1 OAuth1, r *gohttp.Request) string {
+	// create map of all parameters
+	buf, _ := ioutil.ReadAll(r.Body)
+	originalBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	copyBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	r.Body = originalBody
```

Comment:
Not sure what's happening here - you read the body then reassign it back.. and then later we overwrite it with the copy?

Reply:
This was a big problem we ran into and saw other people have online too. When you read out the `Body` of a Request, since it's an `io.ReadCloser`, it dumps the `r.Body` contents; leaving it empty. Even if you `r.Clone()` the body, it will still refer to the same `io.ReadCloser` object and when `r.Body` is read it will dump the contents.

When `r.ParseForm()` is called, the `r.Body` contents are dumped, which causes the overall HTTP request is sent on to the destination without a body. This also causes another error, because the Header `Content-Length: X` is still present, but the actual request body length is now 0.

To fix this we dump out the contents of `r.Body`, put the contents into two separate `io.ReaderCloser` objects, assign one back to `r.Body`(since it is now empty again from being dumped), run through the method(`r.ParseForm()` dumps the body again >:0), and finally assign the second copy to `r.Body` again, so the HTTP Request can be sent on like nothing ever happened.

If you have another idea, please let us know. We worked on this for a while, and it seems like it's a fairly niche, but common issue when ran into.

References:
https://github.com/golang/go/issues/36095
https://stackoverflow.com/questions/23070876/reading-body-of-http-request-without-modifying-request-state
https://stackoverflow.com/questions/28158990/golang-io-ioutil-nopcloser